bokehjs/src/coffee/models/annotations/label.coffee

```diff
@@ -0,0 +1,184 @@
+_ = require "underscore"
+$ = require "jquery"
+
+Annotation = require "./annotation"
+ColumnDataSource = require "../sources/column_data_source"
+Renderer = require "../renderers/renderer"
+p = require "../../core/properties"
+
+class LabelView extends Renderer.View
+  initialize: (options) ->
+    super(options)
+    if not @mget('source')?
+      this.mset('source', new ColumnDataSource.Model())
+    @canvas = @plot_model.get('canvas')
+    @xmapper = @plot_view.frame.get('x_mappers')[@mget("x_range_name")]
+    @ymapper = @plot_view.frame.get('y_mappers')[@mget("y_range_name")]
+    @set_data()
+
+  bind_bokeh_events: () ->
+    if @mget('render_mode') == 'css'
+      # dispatch CSS update immediately
+      @listenTo(@model, 'change', @render)
+    else
+      @listenTo(@model, 'change', @plot_view.request_render)
+
+  set_data: () ->
+    super(@mget('source'))
+    @set_visuals(@mget('source'))
+
+  _set_data: () ->
+    @label_div = ($("<div>").addClass('label').hide() for i in @_text)
+    @width = (null for i in @_text)
+    @height = (null for i in @_text)
+    @x_shift = (null for i in @_text)
+    @y_shift = (null for i in @_text)
+
+    ld = @mget("border_line_dash")
+    if _.isArray(ld)
+      if ld.length < 2
+        @line_dash = "solid"
+      else
+        @line_dash = "dashed"
+    if _.isString(ld)
+        @line_dash = ld
+
+    ctx = @plot_view.canvas_view.ctx
+    for i in [0...@_text.length]
+      @visuals.text.set_vectorize(ctx, i)
+      @width[i] = ctx.measureText(@_text[i]).width
+      @height[i] = ctx.measureText(@_text[i]).ascent / 1.175
+      [ @x_shift[i], @y_shift[i] ] = @_calculate_offset(ctx, @height[i], @width[i])
+
+  _map_data: () ->
+    if @mget('x_units') == "data"
+      vx = @xmapper.v_map_to_target(@_x)
+    else
+      vx = @_x.slice(0) # make deep copy to not mutate
+    sx = @canvas.v_vx_to_sx(vx)
+
+    if @mget('y_units') == "data"
+      vy = @ymapper.v_map_to_target(@_y)
+    else
+      vy = @_y.slice(0) # make deep copy to not mutate
+
+    sy = @canvas.v_vy_to_sy(vy)
+
+    return [sx, sy]
+
+  _calculate_offset: (ctx, height, width) ->
+    if ctx.textAlign == 'left'
+      x_shift = 0
+    if ctx.textAlign == 'center'
+      x_shift = -width / 2
+    if ctx.textAlign == 'right'
+      x_shift = -width
+
+    if ctx.textBaseline == 'top'
+      y_shift = 0.0
+    if ctx.textBaseline == 'middle'
+      y_shift = -0.5 * height
+    if ctx.textBaseline == 'bottom'
+      y_shift = -1.0 * height
+    if ctx.textBaseline == 'alphabetic'
+      y_shift = -0.8 * height
+    if ctx.textBaseline == 'hanging'
+      y_shift = -0.17 * height
```

Comment:
I'll add a link to the w3c spec around textBaseline. The actual baseline are font-specific, so these are guessestimates.